webdriver/tests/element_click/shadow_dom.py

```diff
@@ -0,0 +1,118 @@
+from tests.support.asserts import assert_error, assert_success
+from tests.support.inline import inline
+
+def element_click(session, element):
+    return session.transport.send(
+        "POST", "session/{session_id}/element/{element_id}/click".format(
+            session_id=session.session_id,
+            element_id=element.id))
+
+def get_checkbox_dom():
+    return inline("""
+        <style>
+            custom-checkbox-element {
+                display:block; width:20px; height:20px;
+            }
+        </style>
+        <custom-checkbox-element></custom-checkbox-element>
+        <script>
+            customElements.define('custom-checkbox-element',
+                class extends HTMLElement {
+                    constructor() {
+                            super();
+                            this.attachShadow({mode: 'open'}).innerHTML = `
+                                <div><input type="checkbox"/></div>
+                            `;
+                        }
+                });
+        </script>""")
+
+def test_shadow_element_click(session):
+    session.url = get_checkbox_dom()
+    custom_element = session.find.css("custom-checkbox-element", all=False)
+    is_pre_checked = session.execute_script("return arguments[0].shadowRoot.querySelector('input').checked", args=(custom_element,))
+    assert is_pre_checked == False
+    response = element_click(session, custom_element)
+    assert_success(response)
+    is_post_checked = session.execute_script("return arguments[0].shadowRoot.querySelector('input').checked", args=(custom_element,))
+    assert is_post_checked == True
+
+
+def test_inside_element_click(session):
+    session.url = get_checkbox_dom()
+    custom_element = session.find.css("custom-checkbox-element", all=False)
+    checkbox_element = session.execute_script("return arguments[0].shadowRoot.querySelector('input')", args=(custom_element,))
+    is_pre_checked = session.execute_script("return arguments[0].checked", args=(checkbox_element,))
+    assert is_pre_checked == False
+    click_response = element_click(session, checkbox_element)
+    assert_success(click_response)
+    is_post_checked = session.execute_script("return arguments[0].checked", args=(checkbox_element,))
+    assert is_post_checked == True
```

Comment:
You could do something with [parametrize](https://docs.pytest.org/en/latest/parametrize.html), like `@pytest.mark.parametrize("click_on", ["custom_element", "checkbox_element"])` and then I guess `element_click(session, locals()[click_on])`, which eliminates basically all the repetition. Not sure there's anything much better than that, given we do need to refer to all the elements.

Reply:
Thanks! This cut it down to two tests with parameters.